### PR TITLE
Add cudf::strings:udf::replace function

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Add udf function for replacing all occurrences of a target string within a source string with a replacement string.
This is to support #12207 

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by DaceT** [comment]
run benchmarks

File: python/strings_udf/cpp/include/cudf/strings/udf/replace.cuh [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ * @param source Source string to search
+ * @param target String to match within source
+ * @param replacement String to replace the target within the source
+ * @return Resulting string
+ */
+__device__ inline udf_string replace(string_view source,
+                                     string_view target,
+                                     string_view replacement)
+{
+  if (target.empty()) return udf_string(source);
```
Comment (bdice) [inline python/strings_udf/cpp/include/cudf/strings/udf/replace.cuh:37]: Does this fit Python’s replace logic for empty targets? Or does pandas differ from this behavior / do we have an alternative function?
```python
>>> print('test'.replace('', 'a'))
ataeasata
```

**Review by brandon-b-miller — CHANGES_REQUESTED** [review]
I tried patching https://github.com/rapidsai/cudf/pull/12207 locally with this code and I found that this function passes my tests, _except_ for the cases @bdice mentioned where an empty string is being replaced by a nonempty character or sequence. 

there, my shim function around this is
```c++
extern "C" __device__ int replace(
  int* nb_retval, void* udf_str, void* const src, void* const to_replace, void* const replacement)
{
  auto src_ptr         = reinterpret_cast<cudf::string_view const*>(src);
  auto to_replace_ptr  = reinterpret_cast<cudf::string_view const*>(to_replace);
  auto replacement_ptr = reinterpret_cast<cudf::string_view const*>(replacement);

  auto udf_str_ptr = new (udf_str) udf_string;
  udf_string result;

  result = replace(*src_ptr, *to_replace_ptr, *replacement_ptr);

  *udf_str_ptr = result;

  return 0;
}

**Review by bdice — APPROVED** [review]
LGTM if the tests for empty targets are passing. I suppose the tests aren't supposed to be in this PR, but another one (#12207).

**Review by brandon-b-miller — APPROVED** [review]
Tests pass in my PR now with these changes.

**Comment by davidwendt** [comment]
@gpucibot merge

**Pull request merged**